stdlib/public/core/SmallString.swift

```diff
@@ -127,25 +127,30 @@ extension _SmallUTF8String {
 #endif
   }
 
+  @inline(__always)
   @inlinable
+  @effects(readonly)
   public // @testable
   init?(_ codeUnits: UnsafeBufferPointer<UInt8>) {
 #if arch(i386) || arch(arm)
     return nil // Never form small strings on 32-bit
 #else
     let count = codeUnits.count
     guard count <= _SmallUTF8String.capacity else { return nil }
-    self.init()
-    self._withAllUnsafeMutableBytes { rawBufPtr in
-      let rawDst = rawBufPtr.baseAddress._unsafelyUnwrappedUnchecked
-      memcpy_(
-        dst: rawDst.assumingMemoryBound(to: UInt8.self),
-        src: codeUnits.baseAddress._unsafelyUnwrappedUnchecked,
-        count: count
-      )
+
+    let addr = codeUnits.baseAddress._unsafelyUnwrappedUnchecked
+    var high: UInt
+    let lowCount: Int
+    if count > 8 {
+      lowCount = 8
+      high = _bytesToUInt(addr + 8, count &- 8)
+    } else {
+      lowCount = count
+      high = 0
     }
-    _sanityCheck(self.count == 0, "overwrote count early?")
-    self.count = count
+    high |= (UInt(count) &<< (8*15))
+    let low = _bytesToUInt(addr, lowCount)
+    _storage = (low, high)
```

Comment:
I think the code would be more clear if this was in the if-else blocks. We'd also not need to introduce so many in-scope variables.

Reply:
I wanted to reduce the number of calls to _bytesToUInt. Although llvm can fold away the whole loop, this is not yet done on SIL level, so duplicating _bytes_ToUInt in the then- and else-branch generates an additional loop in SIL.